Including QC module

krischer opened the pull request:
As some of you might know in EIDA we are working on a module to compute QC parameters and extract features from seismic waveforms. Would you be interested in including such a module in obspy?

krischer commented:
This issue is now a pull request to your branch. I'll have a look on monday. Have a nice weekend :-)

marcopovitch commented:
just for your information, there is also an other project that's seem to be active on github, that deals with data quality (I haven't tested it). It is java based :( 

https://github.com/usgs/seedscan

ltrani commented:
Just a quick update on the status... the module is almost ready. I tested and compared the results with the C procedure and they seem to be aligned. More specific tests for edge cases are needed. I will try to produce some simulated scenarios but help there is very welcome. Last issue to solve is the handling of flags crossing the file boundary, e.g.: due to records of the day under analysis which are stored in the previous or next file. At the moment the count is done on a single file but this needs to be changed as it might miss some records.

ltrani commented:
I don't know why the build keeps failing on AppVeyor. Any clue to fix this?

megies commented:
Appveyor has often problem completing our build nowadays, we don't know why yet. Not sure if this the case here, though. There's zillions of `Warning: overflow encountered in int_scalars` lines in the output (https://ci.appveyor.com/project/obspy/obspy/build/1.0.1099-master).. I think this comes from your code..? At least I've never seen this in other builds (e.g. https://ci.appveyor.com/project/obspy/obspy/build/1.0.1098-master). These lines don't appear in Travis build though.. so dunno

ltrani commented:
I see. Actually I'm also wondering about all those warnings. I wasn't sure if it was me but I will have a look.

ltrani commented:
I found this http://lists.swapbytes.de/archives/obspy-users/2015-May/001707.html. It seems that it is not the first time these types of warning appear. They seem to be related to numpy, but I wouldn't know why it is complaining

ltrani commented:
I don't think those warnings come from my code. They don't appear when I run test_mseed_util.py or test_quality_control.py.

megies commented:
> I don't think those warnings come from my code. They don't appear when I run test_mseed_util.py or test_quality_control.py.

I assume you're on Linux.. like I mentioned earlier, we don't see them on Travis, only on Appveyor. Probably has no significance, but unless we get a test result from Windows, it's hard to tell..

ltrani commented:
I'm on mac

megies commented:
Hmm.. OK. Still no idea if this PR works on Windows then..

ThomasLecocq commented:
@megies should I test something ? how to (I know, I'm supposed to know)... ?

megies commented:
> @megies should I test something ? how to (I know, I'm supposed to know)... ?

That would be great!

```bash
conda remove obspy
pip uninstall obspy
pip install https://github.com/ltrani/obspy/archive/mseed-qc.zip
obspy-runtests -r signal io.mseed
```

krischer commented:
The PR looks mostly good. A big issue is the size of the test files. Many of your files are multiple megabytes which unfortunately is not acceptable in a git repository.

Can you please slim-down the test files a lot and rewrite the history of this branch to not include these large files? It should be very possible to keep each test file in the realm of ~10kB. Just use 256 bit mini seed records. I could also assist with both tasks if needed.

krischer commented on obspy/io/mseed/util.py:
```diff
@@ -105,27 +105,40 @@ def get_start_and_end_time(file_or_file_object):
     :param file_or_file_object: Mini-SEED file name or open file-like object
         containing a Mini-SEED record.
-
+    :type str
```
It also already accepts `UTCDateTime` objects and anything the `UTCDateTime` constructor eats. Please document that.

krischer commented on obspy/io/mseed/util.py:
```diff
@@ -238,31 +275,110 @@ def get_timing_and_data_quality(file_or_file_object):
     # Keep track of the extracted information.
     quality_count = [0, 0, 0, 0, 0, 0, 0, 0]
+    activity_count = [0, 0, 0, 0, 0, 0, 0]
+    io_count = [0, 0, 0, 0, 0, 0]
```
We do alreay have a new function here so this is a good opportunity to change the return types. I think a dictionary with flag counts would be more convenient. What do you think?

e.g.

```python
io_count = collections.OrderedDict(
    station_volume_parity_error=0,
    long_record_read=0,
    short_record_read=0, ...)
```

This is easier to read and reason about and thus should reduce usage errors in the long run. The `OrderedDict` means that `io_count.values()` is exactly the same as the current return value.

ltrani replied:
I think it is a good idea but then should we do this for all the counts (quality and activity too)?

krischer replied:
Yes for all counts. The wrapper around the old function can still return the list.

ltrani replied:
I've implemented this feature but it is not as straightforward as I thought to keep the old tests running. The method values() is actually returning something like this odict_values([0, 0, 0, 0, 0, 0, 0, 0]) which is a dictionary view object instead of the simple array. I must say I'm using python 3 and this seems to be causing it. I can solve this but I'm not sure how this would behave with python 2

krischer replied:
You should be able to just do `list(dict_name.values())`.

ltrani replied:
sure but does this work with python 2?

krischer replied:
On Python 2 `.values()` should already return a list. In Python 3 those return iterative view objects. So this should work on both.

The way ObsPy is currently set up causes Python 2 to more or less behave like Python 3 so almost anything that works on Python 3 does also work on Python 2. If not then our tests will catch it.

ltrani replied:
cool

krischer commented on obspy/io/mseed/util.py:
```diff
@@ -238,31 +275,110 @@ def get_timing_and_data_quality(file_or_file_object):
+            # Add the value of each bit to the quality_count.
+            for _i in range(8):
+                    if (this_info['data_quality_flags'] & (1 << _i)) != 0:
+                        quality_count[_i] += 1
```
This is essentially the same thing 4 times in a row. Better do

```python
while offset <= (info['filesize'] - 256):
     this_info = get_record_information(file_or_file_object, offset)
    # Might be a good time to think about and document the exact
    # meaning of the starttime and endtime parameters.
     if starttime is not None an this_info["endtime"] < starttime:
        continue
     if endtime is not None and this_info["starttime"] > endtime:
        continue
     ... extract the flags...
```

ltrani replied:
I agree!

ltrani replied:
The point is that when starttime and endtime are None (default) you should still be able to extract the flags. I didn't want to disrupt the current behaviour.
Starttime and endtime should be optional parameters that users specifie if they want to extract only subsets of the flags related to specific portions of the file(s)

krischer replied:
I agree. That is still be the case with my modified version.

The other question to ask is what happens when you set `starttime` to like 5, and a record goes from 0 to 10. Do we still count that record or only the next one? Same with the end time setting.

ltrani replied:
Good point!  I guess we are somehow forced to accept compromises and approximations intrinsicly associated to the MSEED format. As long this is documented...

ltrani replied:
I can add on this that a proposal has been made to the FDSN WG2... in that proposal the record counts as soon as starttime and/or endtime fall in it, but this applies only to medatata extracted from the fixed header (therefore the once returned by get_flags). Samples metadata are computed on the actual number of samples.

ltrani commented:
To be honest this is kind of tricky for me. I didn't create those test files myself and it was quite challeging to have them right. I had to test special edge cases where records would spread across different days. An alternative would be to create synthetic MSEED files, but I don't think I'm the most appropriate person for that. The point is to check the flags count on multiple files, if you have ideas please ... I think it would be good if someone more domain-savvy had a deeper look at this.

krischer commented:
I can shrink the MiniSEED files. It is easy for the `get_flags` function. I have to try and see how easy it is for the `MSEEDMetadata` class while still testing it enough...

ltrani commented:
>I can shrink the MiniSEED files. It is easy for the get_flags function. I have to try and see how easy it is for the MSEEDMetadata class while still testing it enough... 

Did you manage to do something about this? I fear it is not straight forward. As an alternative could we link external files via url? Or should we rely on a simulated mseed object?

ltrani commented:
```javascript
======================================================================
ERROR: test_get_json_meta_no_tq (obspy.signal.tests.test_quality_control.QualityControlTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "obspy\signal\tests\test_quality_control.py", line 86, in test_get_json_meta_no_tq
    mseed_metadata.populate_metadata(files, start_time, end_time)
  File "obspy\signal\quality_control.py", line 257, in populate_metadata
    self.__read__(files, stime, etime)
  File "obspy\signal\quality_control.py", line 74, in __read__
    self.__st__ += read(file)
  File "<decorator-gen-29>", line 2, in read
  File "obspy\core\util\decorator.py", line 283, in _map_example_filename
    return func(*args, **kwargs)
  File "obspy\core\stream.py", line 232, in read
    st.extend(_read(file, format, headonly, **kwargs).traces)
  File "<decorator-gen-30>", line 2, in _read
  File "obspy\core\util\decorator.py", line 193, in uncompress_file
    result = func(filename, *args, **kwargs)
  File "obspy\core\stream.py", line 276, in _read
    headonly=headonly, **kwargs)
  File "obspy\core\util\base.py", line 452, in _read_from_plugin
    list_obj = read_format(filename, **kwargs)
  File "obspy\io\mseed\core.py", line 371, in _read_mseed
    allocData, diag_print, log_print)
WindowsError: exception: access violation writing 0xFFFFFFFF939119D8
```
Any idea about this?

megies commented:
Maybe related to #1107 but not so likely..
Maybe rebase on recent master before anything else, to exclude possible fixes done in the meantime?

ltrani commented:
Rebased but now I get a number of tests failing (7) like this:

```javascript
======================================================================
FAIL: test_flake8 (obspy.core.tests.test_code_formatting.CodeFormattingTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/trani/git/obspy/obspy/core/tests/test_code_formatting.py", line 29, in test_flake8
    self.assertEqual(error_count, 0, "\n" + message.decode())
AssertionError: 3 != 0 : 
/Users/trani/git/obspy/obspy/io/sac/header.py:211:9: E126 continuation line over-indented for hanging indent
/Users/trani/git/obspy/obspy/io/sac/sactrace.py:1010:25: E126 continuation line over-indented for hanging indent
/Users/trani/git/obspy/obspy/signal/spectral_estimation.py:1437:29: E126 continuation line over-indented for hanging indent



======================================================================
ERROR: test_read_fnetmt_catalog (obspy.io.nied.tests.test_fnetmt_reading.FNETMTCatalogReadingTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/trani/git/obspy/obspy/core/util/base.py", line 437, in _read_from_plugin
    format_ep = EPS[format]
KeyError: 'FNETMT'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/trani/git/obspy/obspy/io/nied/tests/test_fnetmt_reading.py", line 24, in test_read_fnetmt_catalog
    cat = read_events(testfile, 'FNETMT')
  File "<decorator-gen-35>", line 2, in read_events
  File "/Users/trani/git/obspy/obspy/core/util/decorator.py", line 283, in _map_example_filename
    return func(*args, **kwargs)
  File "/Users/trani/git/obspy/obspy/core/event.py", line 141, in read_events
    catalog = _read(pathnames[0], format, **kwargs)
  File "<decorator-gen-36>", line 2, in _read
  File "/Users/trani/git/obspy/obspy/core/util/decorator.py", line 193, in uncompress_file
    result = func(filename, *args, **kwargs)
  File "/Users/trani/git/obspy/obspy/core/event.py", line 154, in _read
    **kwargs)
  File "/Users/trani/git/obspy/obspy/core/util/base.py", line 440, in _read_from_plugin
    raise TypeError(msg % (format, ', '.join(EPS)))
TypeError: Format "FNETMT" is not supported. Supported types: QUAKEML, NDK, CMTSOLUTION, MCHEDR, NLLOC_HYP, ZMAP

======================================================================
```

megies commented:
That's a new IO plugin that got merged into master in the meantime, the list of plugins gets registered with setuptools during install and this step is missing when just updating your linked source-tree installation. So just run the installation on your source tree again. :)

megies commented on obspy/io/mseed/util.py:
```diff
@@ -105,27 +105,40 @@ def get_start_and_end_time(file_or_file_object):
             "'get_timing_and_data_quality'. Use that instead.")
 def getTimingAndDataQuality(*args, **kwargs):
-    return get_timing_and_data_quality(*args, **kwargs)
+    return get_flags(activity_flags=False, io_flags=False, *args, **kwargs)
```
`t_quality=False` per default, nee d to set it `True` here? Also, maybe rename to `timing_quality`?

ltrani replied:
this refers to the full list of values not to the statistics which are returned by default

megies commented:
The errors on Windows are still there in Appveyor, though.. 

But, they are not only appearing in your tests but also in other tests that are just doing a simple read from MSEED (e.g. https://ci.appveyor.com/project/obspy/obspy/build/1.0.1208-master#L1215) and never even execute any of the code you added/changed (as far as I see). So, I don't see how those fails could possibly relate to your changes..

ltrani commented:
>That's a new IO plugin that got merged into master in the meantime, the list of plugins gets registered with setuptools during install and this step is missing when just updating your linked source-tree installation. So just run the installation on your source tree again. :)

thanks but after doing that I get this error when running the tests:

```javascript
Traceback (most recent call last):
  File "runtests.py", line 14, in <module>
    from obspy.scripts.runtests import main
  File "/Users/trani/git/obspy/obspy/__init__.py", line 41, in <module>
    from obspy.core.utcdatetime import UTCDateTime  # NOQA
  File "/Users/trani/git/obspy/obspy/core/__init__.py", line 120, in <module>
    from obspy.core.util import DynamicAttributeImportRerouteModule
  File "/Users/trani/git/obspy/obspy/core/util/__init__.py", line 41, in <module>
    from obspy.core.util.testing import add_doctests, add_unittests
  File "/Users/trani/git/obspy/obspy/core/util/testing.py", line 36, in <module>
    MATPLOTLIB_VERSION = get_matplotlib_version()
  File "/Users/trani/git/obspy/obspy/core/util/base.py", line 328, in get_matplotlib_version
    import matplotlib
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/site-packages/matplotlib/__init__.py", line 1131, in <module>
    rcParams = rc_params()
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/site-packages/matplotlib/__init__.py", line 975, in rc_params
    return rc_params_from_file(fname, fail_on_error)
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/site-packages/matplotlib/__init__.py", line 1100, in rc_params_from_file
    config_from_file = _rc_params_in_file(fname, fail_on_error)
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/site-packages/matplotlib/__init__.py", line 1018, in _rc_params_in_file
    with _open_file_or_url(fname) as fd:
  File "/usr/local/Cellar/python3/3.5.0/Frameworks/Python.framework/Versions/3.5/lib/python3.5/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/site-packages/matplotlib/__init__.py", line 1000, in _open_file_or_url
    encoding = locale.getdefaultlocale()[1]
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/locale.py", line 559, in getdefaultlocale
    return _parse_localename(localename)
  File "/Users/trani/Virtualenvs/eida-py3/lib/python3.5/locale.py", line 487, in _parse_localename
    raise ValueError('unknown locale: %s' % localename)
ValueError: unknown locale: UTF-8
```

megies commented:
Huh.. what does `locale` and `locale -a` say on your machine?

ltrani commented:
>Huh.. what does locale and locale -a say on your machine?

This...
```
locale
LANG=
LC_COLLATE="C"
LC_CTYPE="UTF-8"
LC_MESSAGES="C"
LC_MONETARY="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_ALL=
```
```
locale -a
af_ZA
af_ZA.ISO8859-1
af_ZA.ISO8859-15
af_ZA.UTF-8
am_ET
am_ET.UTF-8
be_BY
be_BY.CP1131
be_BY.CP1251
be_BY.ISO8859-5
be_BY.UTF-8
bg_BG
bg_BG.CP1251
bg_BG.UTF-8
ca_ES
ca_ES.ISO8859-1
ca_ES.ISO8859-15
ca_ES.UTF-8
cs_CZ
cs_CZ.ISO8859-2
cs_CZ.UTF-8
da_DK
da_DK.ISO8859-1
da_DK.ISO8859-15
da_DK.UTF-8
de_AT
de_AT.ISO8859-1
de_AT.ISO8859-15
de_AT.UTF-8
de_CH
de_CH.ISO8859-1
de_CH.ISO8859-15
de_CH.UTF-8
de_DE
de_DE.ISO8859-1
de_DE.ISO8859-15
de_DE.UTF-8
el_GR
el_GR.ISO8859-7
el_GR.UTF-8
en_AU
en_AU.ISO8859-1
en_AU.ISO8859-15
en_AU.US-ASCII
en_AU.UTF-8
en_CA
en_CA.ISO8859-1
en_CA.ISO8859-15
en_CA.US-ASCII
en_CA.UTF-8
en_GB
en_GB.ISO8859-1
en_GB.ISO8859-15
en_GB.US-ASCII
en_GB.UTF-8
en_IE
en_IE.UTF-8
en_NZ
en_NZ.ISO8859-1
en_NZ.ISO8859-15
en_NZ.US-ASCII
en_NZ.UTF-8
en_US
en_US.ISO8859-1
en_US.ISO8859-15
en_US.US-ASCII
en_US.UTF-8
es_ES
es_ES.ISO8859-1
es_ES.ISO8859-15
es_ES.UTF-8
et_EE
et_EE.ISO8859-15
et_EE.UTF-8
eu_ES
eu_ES.ISO8859-1
eu_ES.ISO8859-15
eu_ES.UTF-8
fi_FI
fi_FI.ISO8859-1
fi_FI.ISO8859-15
fi_FI.UTF-8
fr_BE
fr_BE.ISO8859-1
fr_BE.ISO8859-15
fr_BE.UTF-8
fr_CA
fr_CA.ISO8859-1
fr_CA.ISO8859-15
fr_CA.UTF-8
fr_CH
fr_CH.ISO8859-1
fr_CH.ISO8859-15
fr_CH.UTF-8
fr_FR
fr_FR.ISO8859-1
fr_FR.ISO8859-15
fr_FR.UTF-8
he_IL
he_IL.UTF-8
hi_IN.ISCII-DEV
hr_HR
hr_HR.ISO8859-2
hr_HR.UTF-8
hu_HU
hu_HU.ISO8859-2
hu_HU.UTF-8
hy_AM
hy_AM.ARMSCII-8
hy_AM.UTF-8
is_IS
is_IS.ISO8859-1
is_IS.ISO8859-15
is_IS.UTF-8
it_CH
it_CH.ISO8859-1
it_CH.ISO8859-15
it_CH.UTF-8
it_IT
it_IT.ISO8859-1
it_IT.ISO8859-15
it_IT.UTF-8
ja_JP
ja_JP.eucJP
ja_JP.SJIS
ja_JP.UTF-8
kk_KZ
kk_KZ.PT154
kk_KZ.UTF-8
ko_KR
ko_KR.CP949
ko_KR.eucKR
ko_KR.UTF-8
lt_LT
lt_LT.ISO8859-13
lt_LT.ISO8859-4
lt_LT.UTF-8
nl_BE
nl_BE.ISO8859-1
nl_BE.ISO8859-15
nl_BE.UTF-8
nl_NL
nl_NL.ISO8859-1
nl_NL.ISO8859-15
nl_NL.UTF-8
no_NO
no_NO.ISO8859-1
no_NO.ISO8859-15
no_NO.UTF-8
pl_PL
pl_PL.ISO8859-2
pl_PL.UTF-8
pt_BR
pt_BR.ISO8859-1
pt_BR.UTF-8
pt_PT
pt_PT.ISO8859-1
pt_PT.ISO8859-15
pt_PT.UTF-8
ro_RO
ro_RO.ISO8859-2
ro_RO.UTF-8
ru_RU
ru_RU.CP1251
ru_RU.CP866
ru_RU.ISO8859-5
ru_RU.KOI8-R
ru_RU.UTF-8
sk_SK
sk_SK.ISO8859-2
sk_SK.UTF-8
sl_SI
sl_SI.ISO8859-2
sl_SI.UTF-8
sr_YU
sr_YU.ISO8859-2
sr_YU.ISO8859-5
sr_YU.UTF-8
sv_SE
sv_SE.ISO8859-1
sv_SE.ISO8859-15
sv_SE.UTF-8
tr_TR
tr_TR.ISO8859-9
tr_TR.UTF-8
uk_UA
uk_UA.ISO8859-5
uk_UA.KOI8-U
uk_UA.UTF-8
zh_CN
zh_CN.eucCN
zh_CN.GB18030
zh_CN.GB2312
zh_CN.GBK
zh_CN.UTF-8
zh_HK
zh_HK.Big5HKSCS
zh_HK.UTF-8
zh_TW
zh_TW.Big5
zh_TW.UTF-8
C
POSIX

```

megies commented:
I think these locale settings are just wrong.. see #1132. You probably want `en_US.UTF-8` in most of those (or some italian utf-8 if your OS is in italian).

ltrani commented:
I don't understand because if I test everything seems ok.

```javascript
In [1]: import locale

In [2]: locale.getpreferredencoding()
Out[2]: 'UTF-8'
```

megies commented:
Just `UTF-8` seems to be no valid locale..?
![screenshot from 2015-11-26 15 56 11](https://cloud.githubusercontent.com/assets/1842780/11425593/3fdc61c4-9456-11e5-8750-e769880e5799.png)

ltrani commented:
Ok this is solved but now I get something worse... apparently the new matplotlib version makes stricter checks on the underline backend
```
RuntimeError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are Working with Matplotlib in a virtual enviroment see 'Working with Matplotlib in Virtual environments' in the Matplotlib FAQ
```

and unfortunately this workaround (http://matplotlib.org/1.5.0/faq/virtualenv_faq.html) doens't seem to work...

megies commented:
Any chance Anaconda is an option for you?

ltrani commented:
Done but after rebasing I get plenty of errors generated by other modules. There are even formatting issues like this:
```
======================================================================
FAIL: test_flake8 (obspy.core.tests.test_code_formatting.CodeFormattingTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/trani/git/obspy/obspy/core/tests/test_code_formatting.py", line 29, in test_flake8
    self.assertEqual(error_count, 0, "\n" + message.decode())
AssertionError: 3 != 0 : 
/Users/trani/git/obspy/obspy/io/sac/header.py:211:9: E126 continuation line over-indented for hanging indent
/Users/trani/git/obspy/obspy/io/sac/sactrace.py:1010:25: E126 continuation line over-indented for hanging indent
/Users/trani/git/obspy/obspy/signal/spectral_estimation.py:1437:29: E126 continuation line over-indented for hanging indent
```

ltrani commented:
Other problems look like this: 
```
======================================================================
FAIL: test_remove_response_plot (obspy.core.tests.test_trace.TraceTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/trani/git/obspy/obspy/core/tests/test_trace.py", line 2159, in test_remove_response_plot
    water_level=60, end_stage=None, plot=ic.name)
  File "/Users/trani/git/obspy/obspy/core/util/testing.py", line 405, in __exit__
    raise ImageComparisonException(msg)
obspy.core.util.testing.ImageComparisonException: Image comparision failed.
	RMS:       24.50307595851124
	Tolerance: 3.0
Upload to imgur failed (caught ImportError: No module named 'urlparse').
```

I'm using python 3 where ```urlparse``` has been replaced by ```urrlib.parse```

megies commented:
Don't worry too much about local errors (esp. in unrelated modules) they can have plenty reasons (in the above case your pyimgur is not py3 compatible, see Damgaard/PyImgur#25, reinstall pyimgur from master..), just update this branch here and we can rely on CI for definitive results.

ltrani commented:
I tried all the suggested steps but unfortunately there are still problems with the AppVeyor build. From what I can see this has nothing to do with my code but it is rather related to ```obspy\io\mseed\core.py```
```
File "obspy\io\mseed\core.py", line 371, in _read_mseed
    allocData, diag_print, log_print)
```

On some forums I read this might be related to a C compiler configuration... I don't know how to proceed

megies commented:
You're not changing C codes, as far as I can see. Neither are you using any C routines of libmseed. So I don't see how this can be due to compiler configuration. If it would fail on Travis, too, I would say your changes might affect the position of file pointers that later go to C again, but since it works on Travis.. no idea.
But I'm not too deep into the `obspy.io.mseed` insides, maybe @krischer has an idea what's going on..

petrrr commented:
@ltrani: The test `test_quality_control.QualityControlTestCase` is a bit verbose, is there some `print` or similar left over?

```python
test_populate_metadata_multiple_files (obspy.signal.tests.test_quality_control.QualityControlTestCase) ... defaultdict(None, {'files': ['/Users/petr/sandbox/projects/obspy/obspy/signal/tests/data/NA.SEUT..BHZ.D.2015.289', '/Users/petr/sandbox/projects/obspy/obspy/signal/tests/data/NA.SEUT..BHZ.D.2015.290'], 'event_in_progress': 0, 'sample_max': 9289, 'sample_min': -4580, 'missing_padded_data': 0, 'cha': 'BHZ', 'calibration_signal': 0, 'telemetry_sync_error': 0, 'timing_quality_mean': 99.87734386014381, 'encoding': 'STEIM2', 'loc': '', 'end_time': '2015-10-17T23:57:49.500000', 'amplifier_saturation': 0, 'clock_locked': 7087, 'num_records': 238242, 'sample_rate': 40.0, 'num_samples': 1688501, 'glitches': 0, 'sample_mean': 4062.0917470584855, 'overlaps_len': 0, 'gaps_len': 44127.5, 'sample_stdev': 1983.4446454276149, 'percent_availability': 48.89101227704424, 'event_end': 0, 'timing_quality_min': 80.0, 'quality': 'D', 'suspect_time_tag': 6, 'timing_quality_max': 100.0, 'num_overlaps': 0, 'sta': 'SEUT', 'net': 'NA', 'sample_rms': 4520.4692259760086, 'spikes': 0, 'record_len': 512, 'event_begin': 0, 'digitizer_clipping': 0, 'start_time': '2015-10-17T00:00:00', 'timing_correction': 0, 'digital_filter_charging': 0, 'num_gaps': 32})
ok
```

ltrani commented:
yes but I don't think that's the issue. Anyway I commented it

petrrr commented:
@ltrani: Agreed! this is not related to the failing tests, just a cleanup.

ltrani commented:
Hi guys any updates about the windows build failure?

krischer commented:
Something really weird is going on with this branch - I just went through a very painful rebase but now all is good. Maybe this fixes the windows build issues. @ltrani Can you add me as a collaborator to your ObsPy fork? Then I can force push and fix this branch.

ltrani commented:
Done

krischer commented:
> If you're rebasing, you might as well delete this file when it was first added.

Good point. I'll have to rebase once again to remove the large files - I'll do it then.

krischer commented:
I changed all tests so they generate their own files on the fly thus we don't need the big test files anymore. I also rewrote the history to remove the large files - please remember to checkout the whole tree and not merge this in your potentially existing branch!

While at it I made a number of other changes:

* The `obspy.io.mseed.util.get_flags()` method now returns dictionaries with the flag counts for each flag. I think this is much clearer compared to the old way of just returning a list:

```python
In [5]: get_flags("./qualityflags.mseed")
Out[5]:
{'activity_flags': OrderedDict([('calibration_signals_present', 0),
              ('time_correction_applied', 0),
              ('beginning_event', 0),
              ('end_event', 0),
              ('positive_leap', 0),
              ('negative_leap', 0),
              ('event_in_progress', 0)]),
 'data_quality_flags': OrderedDict([('amplifier_saturation_detected', 9),
              ('digitizer_clipping_detected', 8),
              ('spikes_detected', 7),
              ('glitches_detected', 6),
              ('missing_data_present', 5),
              ('telemetry_sync_error', 4),
              ('digital_filter_charging', 3),
              ('time_tag_uncertain', 2)]),
 'io_and_clock_flags': OrderedDict([('station_volume_parity_error', 0),
              ('long_record_read', 0),
              ('short_record_read', 0),
              ('start_time_series', 0),
              ('end_time_series', 0),
              ('clock_locked', 0)]),
 'record_count': 18,
 'timing_quality': {}}
```

* I added a bunch of tests to `obspy.signal.quality_control`.
* I also internally restructured the quality control module to make it easier to use and reason about. Usage is now like this:

```python
In [12]: MSEEDMetadata(files=["qualityflags.mseed"], c_seg=False).meta
Out[12]:
{'amplifier_saturation': 9,
 'calibration_signal': 0,
 'cha': 'EHE',
 'clock_locked': 0,
 'digital_filter_charging': 3,
 'digitizer_clipping': 8,
 'encoding': ['STEIM1'],
 'end_time': 2008-01-01T00:00:01.970000Z,
 'event_begin': 0,
 'event_end': 0,
 'event_in_progress': 0,
 'files': ['qualityflags.mseed'],
 'gaps_len': 0.0,
 'glitches': 6,
 'loc': '',
 'missing_padded_data': 5,
 'net': 'BW',
 'num_gaps': 0.0,
 'num_overlaps': 17,
 'num_records': 324,
 'num_samples': 7416,
 'overlaps_len': 34.93500113487244,
 'percent_availability': 100.0,
 'quality': 'D',
 'record_len': [512],
 'sample_max': -353,
 'sample_mean': -402.45873786407765,
 'sample_min': -475,
 'sample_rate': [200.0],
 'sample_rms': 4.6786837745071406,
 'sample_stdev': 19.07384083587403,
 'spikes': 7,
 'sta': 'BGLD',
 'start_time': 2007-12-31T23:59:59.915000Z,
 'suspect_time_tag': 2,
 'telemetry_sync_error': 4,
 'timing_correction': 0,
 'timing_quality_lower_quartile': None,
 'timing_quality_max': None,
 'timing_quality_mean': None,
 'timing_quality_median': None,
 'timing_quality_min': None,
 'timing_quality_upper_quartile': None}
```

This is also the structure of the JSON string.

@ltrani Is the structure for this JSON string still open for discussion? So far I took the liberty to change two small things:

* `"encoding"`, `"record_len"`, and `"sample_rate"` are now lists instead of single values as they might potentially change for example if the digitizer is changed or reconfigured.
* I added `"timing_quality_median"`, `"timing_quality_lower_quartile"`, and `"timing_quality_upper_quartile"` fields as these are very useful to know.

I would like to change a couple more things:

* Group all flags in the fixed MiniSEED header and also add all of them, e.g.:
```json
...
"miniseed_header_flag_counts": {
    "data_quality_flags": {
        "amplifier_saturation_detected": 25,
        "digitizer_clipping_detected": 12,
        "spikes_detected": 30,
        "glitches_detected": 6,
        "missing_data_present": 15,
        "telemetry_sync_error": 16,
        "digital_filter_charging": 4,
        "time_tag_uncertain": 8},
    "activity_flags": {
        "calibration_signals_present": 10,
        "time_correction_applied": 20,
        "beginning_event": 33,
        "end_event": 33,
        "positive_leap": 34,
        "negative_leap": 10,
        "event_in_progress": 15},
    "io_and_clock_flags": {
        "station_volume_parity_error": 8,
        "long_record_read": 33,
        "short_record_read": 24,
        "start_time_series": 31,
        "end_time_series": 24,
        "clock_locked": 32}
    }
...
```

* Have more verbose names, .e.g. `"network"` instead of `"net"`, `"gaps_length_in_s"` instead of `"gaps_len"`, ...


Additionally we could also think about enabling the quality metric extraction into our existing plug-in system so people could define their own metrics.

krischer commented:
CI passes now. I am not sure what the previous AppVeyor issue was about but it was related to reading MiniSEED files. Maybe some weird compiler settings on AppVeyor? I'm fairly certain MiniSEED reading otherwise works on Windows - otherwise we would quickly get lots of complaints. Thus I think we can ignore this for now.

ltrani commented:
Great job thanks @krischer ! 

>The ```obspy.io.mseed.util.get_flags()``` method now returns dictionaries with the flag counts for each flag. 

Perfect! I had changed the internal structure using ```OrderedDict``` but I was still returning the list as I was not sure whether you wanted to change the existing method signature

> Is the structure for this JSON string still open for discussion? 

Yes it is and actually I find your suggestions very valid!
I will include these points in the specs we are exchanging on the EIDA mailing list in order to have them in the standard API under definition. 

> Group all flags in the fixed MiniSEED header and also add all of them...

I like very much this idea.

>Additionally we could also think about enabling the quality metric extraction into our existing plug-in system so people could define their own metrics.

I think this is a great idea with a lot of potential for future applications

krischer commented:
Great. I propose to also provide a JSONSchema (http://spacetelescope.github.io/understanding-json-schema/) for the definition to enforce correctness. Also please include a `version` field in the JSON file.

ltrani commented:
We already have a JSON schema but a ```version``` field is a good idea

ltrani commented:
Hi, someone knows what is happening with this branch? I thought the failure of the tests was due to the recent pull request but reverting it does not fix the problem. 
Thanks

megies commented:
> Hi, someone knows what is happening with this branch? I thought the failure of the tests was due to the recent pull request but reverting it does not fix the problem.

There seem to be some merge conflicts with current master (maybe only trivial ones, not sure). It would be best if you rebase this branch on the current master.

One thing.. looking at the `signal/quality_control.py` module.. it seems to be almost exclusively MiniSEED centric, which is OK because that's what data centers use.
And, it's strictly speaking no signal processing but rather (MiniSEED specific) `file introspection`, right..?
So, maybe it should be moved over to `io/mseed` module?

megies commented:
If you have problems with it.. I just rebased your branch, you can check it out here: https://github.com/megies/obspy/tree/mseed-qc-rebase

ltrani commented:
Hi @megies thank you indeed it is probably better to have it under io/mseed. Actually we were thinking whether it would make sense to have this as a property of the stream object. What is your view?

megies commented:
> Actually we were thinking whether it would make sense to have this as a property of the stream object. What is your view?

Hmm.. It's not clear for me how/what you exactly want to attach to the stream (I`m not 100% familiar with all details of the PR) but I have the impression this is interesting to a relatively small part of the user base, so I'm not sure if we should add it to Stream (to not clutter Stream's namespace with specialty routines). Only my view, though.. @krischer?

Jollyfant commented:
The Stream methods `Stream.std()` and `Stream.min()` can then be collected within the more general Stream.quality() method and we actually reduce the namespace clutter. The QC parameters are calculated for a stream object so it would make sense to attach it. It would also give quality parameters that are useful for the entire user base if we could include a flag that defaults to *simple* and returns only `min`, `max`, `rms`, `std`, `ngaps`. Then for the datacenters the other less common QC parameters can be returned too by setting this flag to *full*.

megies commented:
Something utterly went wrong with your rebase..

If you didn't add any new changes yesterday, I'd recommend forcing the tip of this branch to the result of my rebase from yesterday (I can do it for you if you give me push access to your fork):

> @ltrani If you have problems with it.. I just rebased your branch, you can check it out here: https://github.com/megies/obspy/tree/mseed-qc-rebase

megies commented:
> The Stream methods Stream.std() and Stream.min() can then be collected within the more general Stream.quality() method and we actually reduce the namespace clutter.

Sounds fair enough..

ltrani commented:
Thank you @megies indeed I had some problems rebasing it... your help is very welcome I added you as collaborator

megies commented:
Force-pushed my rebased version.

claudiodsf commented:
Just a comment: on one side we are saying that this is really mseed-centric; on the other side we are proposing to attach this to a generic container like the `Stream()` object. 
If we choose to do so, we should be sure that `Stream.quality()` works with any data format.

megies commented:
> Just a comment: on one side we are saying that this is really mseed-centric; on the other side we are proposing to attach this to a generic container like the Stream() object.
> If we choose to do so, we should be sure that Stream.quality() works with any data format.

Yes, indeed. But as I understood, the addition to `Stream` is about simple data metrics, non-specific to mseed.

Jollyfant commented:
Thanks @claudiodsf, now I'm convinced this class is too dedicated to MSEED files and should maybe be moved to io.mseed like @megies suggested. But it is still be a good idea to make a `.quality()` method that works for all streams and collects the present `.std()` and `.min()` methods.

Jollyfant commented:
@megies, any idea why AppVeyor is failing on basemap?

megies commented:
> @megies, any idea why AppVeyor is failing on basemap?

This should be fixed with #1292, you can just wait until it is merged and then rebase.

Jollyfant commented:
@krischer, are you still working on this branch? Currently I'm restructuring the output as you suggested here https://github.com/obspy/obspy/pull/1141#issuecomment-174478868, in addition to adding miniseed_header_flag_percentages instead of just counts. Checking so we don't conflict later.

krischer commented:
Not right now. I'm planning to have another looks in a week or two but I'll let you know beforehand. Feel free to make any chance!

megies commented:
> This should be fixed with #1292, you can just wait until it is merged and then rebase.

Btw, if you rebase on current master, the appveyor/basemap problem should be gone.

Jollyfant commented:
So I rebased my copy of this branch on to current master. Little bit wary of force pushing this to @ltrani's branch in case I screwed it up somehow. And since I'll probably be continuing work on this branch instead of @ltrani anyway perhaps it's a good idea to open a new pull request from my fork and continue the development there.

Idk what's the best way to handle this, what do you guys think? This branch will probably need a lot of work (based on user feedback) before finally being merged in to master.

megies commented:
> Little bit wary of force pushing this to @ltrani's branch in case I screwed it up somehow.

Git doesn't lose anything, ever (unless you're deliberately reworking the internal history -- [like we did when migrating to git ;-)](https://github.com/obspy/obspy/wiki/Svn-to-git-migration)). Actually it'd be enough to write down the current commit hash of `ltrani:mseed-qc`'s branch on a piece of paper before force-pushing to it. ;-) You could always `git reset --hard <HASH>` any branch to represent the exact same history afterwards.
Usually what I do before force-pushing to a branch when unsure is to just do a "backup copy" of the branch with e.g. `git branch mseed-qc_BACKUP mseed-qc`. You can easily restore it from there.. (going to that backup branch and finding out the commit hash of it's tip)

> perhaps it's a good idea to open a new pull request from my fork and continue the development there.

We really prefer to keep development on one thing on the very same ticket, otherwise it's much harder to follow discussions. Also, it's decreasing the danger of people keeping to develop on diverging histories without noticing (always good to push commits as soon as possible when working on PRs jointly).

Short story: Do some local backup branch (can also push the backup somewhere as well) and force-push to update the branch of this PR.

Jollyfant commented:
Ok thanks Tobias. I made another attempt to rebase to the head of master and forced it to @ltrani. Let's see if it goes well..

Jollyfant commented:
@megies, do you know why the code coverage bird is failing us? I don't understand what it means talking about `core/util/version.py`.. something we haven't modified.

megies commented:
> @megies, do you know why the code coverage bird is failing us? I don't understand what it means talking about core/util/version.py.. something we haven't modified.

Don't worry, coverage sometimes shows strange warnings on files not part of PRs. No idea why. Coverage and CI results otherwise look just fine, so just ignore.

I would like to wait for @krischer's review (out of office currently) if you're in no hurry to move on with this.

Also note that we're now building docs for PRs automatically, build is underway and should be ready in maybe half an hour (and should pop up as a new build status notification at the bottom). So you can review your docs changes there, to see if they render as expected.

Jollyfant commented:
I agree it is good idea to wait for code review from @krischer before merging. The automatic building of docs is a nice feature, good job on that. There are some problems in the doc that I will fix tomorrow.

Jollyfant commented:
Currently we're using the Stream.get_gaps() method to calculate gaps and overlaps in a given stream, but this method only compares end times & start times of consecutive traces. It does not calculate gaps and overlaps in the way we need it for this module so I'm working on a better routine for this right now.

Example of the problematic stream:

    IU.GRFO..HH2 | 2016-01-07T00:00:00.008300Z - 2016-01-07T00:00:30.098300Z | 100.0 Hz, 3010 samples
    IU.GRFO..HH2 | 2016-01-07T00:00:02.668393Z - 2016-01-07T00:00:09.518393Z | 100.0 Hz, 686 samples
    IU.GRFO..HH2 | 2016-01-07T00:00:09.528393Z - 2016-01-07T00:00:50.378393Z | 100.0 Hz, 4086 samples
    IU.GRFO..HH2 | 2016-01-07T00:00:30.108393Z - 2016-01-07T00:00:50.378393Z | 100.0 Hz, 2028 samples
    IU.GRFO..HH2 | 2016-01-07T00:00:50.388393Z - 2016-01-07T00:01:51.958393Z | 100.0 Hz, 6158 samples
    IU.GRFO..HH2 | 2016-01-07T00:00:57.248393Z - 2016-01-07T00:01:10.938393Z | 100.0 Hz, 1370 samples
    IU.GRFO..HH2 | 2016-01-07T00:01:31.458393Z - 2016-01-07T00:02:12.518393Z | 100.0 Hz, 4107 samples
    IU.GRFO..HH2 | 2016-01-07T00:01:51.968393Z - 2016-01-07T00:02:12.518393Z | 100.0 Hz, 2056 samples

Gaps:

    IU.GRFO..HH2      2016-01-07T00:00:30.098300Z 2016-01-07T00:00:02.668393Z -6.850000       -685    
    IU.GRFO..HH2      2016-01-07T00:00:50.378393Z 2016-01-07T00:00:30.108393Z -20.270000      -2027   
    IU.GRFO..HH2      2016-01-07T00:01:51.958393Z 2016-01-07T00:00:57.248393Z -13.690000      -1369   
    IU.GRFO..HH2      2016-01-07T00:01:10.938393Z 2016-01-07T00:01:31.458393Z 20.510000       2051    
    IU.GRFO..HH2      2016-01-07T00:02:12.518393Z 2016-01-07T00:01:51.968393Z -20.550000      -2055

The routine identifies a gap (4th gap entry) that is in reality covered by another trace earlier in the stream (5th trace entry) and shouldn't show up as a gap in the quality control.

megies commented:
I would consider this a bug in `Stream.get_gaps()` that should ideally be fixed. The same functionality is in [obspy-scan](https://github.com/obspy/obspy/blob/master/obspy/imaging/scripts/scan.py) and it should be correct there. It might need some refactoring to make it reusable for your use case though..

Jollyfant commented:
obspy-scan seems to give the same result as `Stream.get_gaps()` and is also not a solution (I checked the percentage availabilities of both methods). I have a new little routine that hopefully solves the problem.. which I'll push to this branch in a minute.

megies commented:
> obspy-scan seems to give the same result as Stream.get_gaps()

hmm.. indeed. should be fixed as well.. ;-)

Jollyfant commented:
The performance of the get_flags routine (reverted to the latest version Kirscher worked on) was a little disappointing and I added get_flags_c that is based on clibmseed. This function gets header percentages instead of counts and does not allow overlapping records to be counted more than once.

One question for the experienced:

I'm reading the sampling rate from the fixed section `msr.contents.samprate` but this could prove problematic if the value is not set? I tried using `clibmseed.msr_samprate(msr)` but do not understand the return of this function.

Thanks!

Jollyfant commented:
Pushed some more changes to the continuous segments. Data loaded from different files were put in different traces despite being continuous. It should now concatenate the data of all continuous traces before doing the metric calculation.

Also fixed test where segment length now extends to `endtime + delta`.

Jollyfant commented:
Test was added for a continuous segment from 2 files. I'm pretty satisfied with the state, coverage, and quality of tests for this branch now. So I'm thinking its far enough to merge this in to `obspy.mseed-qc` instead of `master`, so it is included with obspy for now. Until @krischer has had time to review it, and we can merge it to master.

petrrr commented:
@ltrani @Jollyfant 
We started now with doing some proper testing and looking into some details. 

My first observation are somewhat cosmetic and would regard some naming details:

 * The procedure parameter `c_seg` is somewhat compact/cryptic. I'd propose to make it somewhat more "Pythonic", for example `add_c_segments`);
 * I'd also propose to rename `seg_len` to `segment_length`, for consistency. You use `c_segments` and `record_length` elsewhere;
 * I'd propose to rename `sample_rate` in `sampling_rate`. This attribute has a fundamentally different semantic wrt to the other `sample_*` attributes and in `obspy.core.trace.Stats` we use the latter attribute name for exactly the same purpose;
 * I think that `start_time` .. `end_time` on the c_segment has a different meaning wrt to the of the metadata object, where this defines the start and end time as requested on construction, while `first_sample` .. `last_sample` gives the start / end time as found in data. I therefore propose to use the latter also on the c_segment level instead.

megies commented:
+DOCS  (but buildbot is currently incapacitated by the combination of conda/conda#2826 and conda/conda#2699, so nothing I can do until a fixed conda version comes out..)

Jollyfant commented:
@petrrr Thanks for the feedback.

I agree and have applied the changes for points 1, 2, and 3. For point 4 I have renamed `start_time` and `end_time` for the continuous segments to `segment_start` and `segment_end` respectively. Because the end time of a segment is not always the last sample, but Δt further, and may be cut off by the user specified window. This naming convention seems OK to me.

Jollyfant commented:
Ok good idea. I will be adding a small example to the class docstring. @megies is your doc builder bot still alive?

ThomasLecocq commented:
+DOCS

megies commented:
> @megies is your doc builder bot still alive?

Hmm, yeah.. not sure why it didn't pick up your recent commits. I've queued a build manually..

megies commented:
@Jollyfant @ThomasLecocq docs build is up: http://docs.obspy.org/pull-requests/1141/

megies commented:
> Hmm, yeah.. not sure why it didn't pick up your recent commits. I've queued a build manually..

Turns out there were >100 comments before the `+DOCS`, so the magic trigger wasn't on the first page of comments.. to avoid potential confusion in the future, I've added pagination in that check, see da3a3e779bbcb1d7f6e69f21b7f09eb101534524.

Also, I've linked the log of the PR docs bot, http://docs.obspy.org/pull-requests/pull_request_docs.log.

Jollyfant commented:
Sneaky. Good work though, and the docs/example are working and seem OK.

ThomasLecocq commented:
1/ Is that normal that http://docs.obspy.org/pull-requests/1141/packages/autogen/obspy.signal.quality_control.MSEEDMetadata.html#obspy.signal.quality_control.MSEEDMetadata doesn't show the definition of the parameters ? 

2/ And in the example, it would be good to mention files have to have the same ID. 

3/ In my case, the comparison of the trace ids was bad because I don't have a quality flag in the miniseed files... any hint (aside of rewriting all my miniseed files).

megies commented:
> 1/ Is that normal that http://docs.obspy.org/pull-requests/1141/packages/autogen/obspy.signal.quality_control.MSEEDMetadata.html#obspy.signal.quality_control.MSEEDMetadata doesn't show the definition of the parameters ?

They're hidden on the `__init__()` method page: http://docs.obspy.org/pull-requests/1141/packages/autogen/obspy.signal.quality_control.MSEEDMetadata.__init__.html#obspy.signal.quality_control.MSEEDMetadata.__init__
That's a bit annoying and therefore we usually move them from `def __init__(...):` docstring up to the `class XXX:` docstring. Maybe @Jollyfant should do the same here.

Jollyfant commented:
Yeah, fair enough. I moved the parameter definitions to the main class docstring.

> 2/ And in the example, it would be good to mention files have to have the same ID. 

Added this information to the docs.

> 3/ In my case, the comparison of the trace ids was bad because I don't have a quality flag in the miniseed files... any hint (aside of rewriting all my miniseed files).

In your case, I think it would be easiest to comment out the ValueError that is raised when it detects multiple stream ids at line 141 in ``obspy.signal.quality_control.py``. I'd like to keep the exception in place.

ThomasLecocq commented:
>In your case, I think it would be easiest to comment out the ValueError that is raised when it detects multiple stream ids at line 141 in obspy.signal.quality_control.py. I'd like to keep the exception in place.

are quality flags compulsory in the MiniSEED norm ?

Jollyfant commented:
> are quality flags compulsory in the MiniSEED norm ?

I wouldn't say compulsory as much as it is good practise. 

Does in your case ``Trace.stats.mseed.dataquality`` appear as ``None`` and raise an error? We could maybe skip the quality indicator check if it is completely unset, and raise when it is set and does not match.

ThomasLecocq commented:
yes, it's `None`

Jollyfant commented:
Now I'm converting ``None`` to 'none' so it shouldn't raise an error. It still won't let you mix different quality indicators, so all files must be unset for it to work.

megies commented on obspy/signal/quality_control.py:
```diff
@@ -138,8 +138,8 @@ def __init__(self, files, starttime=None, endtime=None,
-        ids = set(tr.id + "." + tr.stats.mseed.dataquality for tr in self.data)
-        if len(ids) != 1:
+        ids = [t.id + "." + str(t.stats.mseed.dataquality) for t in self.data]
+        if len(set(ids)) != 1:
```
@Jollyfant please use `tr` for traces in general whenever possible. If there is a name clash with another local variable called `tr` you can use e.g. `tr_` in the list comprehension.

Jollyfant replied:
Nah, no conflicts. I just didn't want it to pass 80 characters but I'll revert it.

ThomasLecocq commented:
good to merge ?

Jollyfant commented:
The code has been stable for some months and we've used it to ingest metrics for roughly 10TB of data for a test deployment. I think @krischer still wants to do a code review but he's been busy lately. IMHO it is good to merge but I'd rather wait for confirmation from @krischer.

krischer commented:
Sorry for the delay! I'll do it today!

krischer commented:
This took a bit longer then expected - sorry about that. As this is such a long pull request I just directly made some changes here: https://github.com/ltrani/obspy/pull/11 Please review and merge if appropriate.

It simplifies a couple of things and speeds it up a bit.

It would be nice to use `Stream.get_gaps()` for the `._get_gaps_and_overlaps()` method and thus not having to read the data twice but this would be surprisingly hard to implement for a number of reasons.

One thing (aside from my changes) that would be nice to have is to add a test that actually tests the produced JSON against the JSON schema. Is that finalized yet? It could be implemented as an optional test that is only run if the `jsonschema` module is available (most people will have it in any case as the jupyter notebooks depend on it).

Otherwise this looks really really good and test coverage is excellent and comprehensive.

Jollyfant commented:
Thanks Krischer. Changing all the flag names must have been a pain but everything looks good and still works with the ingestion procedures. So I will merge your pull to @ltrani asap.

We were using `Stream.get_gaps()` before but was giving incorrect results for some files (see #1319, #1403). But using `headonly` for the gap and overlap calculation was a smart move and should save some time.

The schema is finalized and will add it as signal/tests/data/wf_metadata_schema.json.

Jollyfant commented:
Rebased to current master and force pushed.

krischer commented:
Turns out I have push-rights to this fork :-)

I made some more changes - two changes to the schema itself - please review!
* c7eeefb6d1f522ede1e88779fd6f0ed352364a7c: Lots of whitespace changes - the actual changes are the changes to the enum values - the schema was not an actually valid JSON schema before this.
* cbc01da32ab6ed0ef5f4d516355328249c0e38c4: Reduced the required nesting for the `producer` key. Otherwise it is a bit awkward.

The `.get_json_meta()` method now has an optional `validate` kwarg. Settings this to `True` will validate the QC JSON string against the schema before returning it.


I also had to add some more meta data and change the name of a view keys to be able to validate the returned dictionary against the schema. Again, please review this commit: https://github.com/obspy/obspy/pull/1141/commits/31211b4054193f9ce065feb561c601d190474b80

Can this be considered a finalized version `1.0.0` of the schema? Once we merge this any change to the schema will require a version number increment.

krischer commented:
> We were using Stream.get_gaps() before but was giving incorrect results for some files (see #1319, #1403).

The issue I got stuck with were the start and end gaps for the cases where `starttime` and `endtime` are given during the metrics calculation. This would trim the files potentially resulting in a tiny gap at the beginning or the end but this should not be counted as a gap as there actually is data...oh well...the current implementation is well tested and seems to cover all the edge cases :-)

Jollyfant commented:
Commit https://github.com/obspy/obspy/pull/1141/commits/31211b4054193f9ce065feb561c601d190474b80 undoes some naming conventions requested by @petrrr. e.g. `sample_rate` to `sampling_rate` but it is kinda arbitrary what we decide roll with. If this matches the schema, let's stick with that. 

> Can this be considered a finalized version 1.0.0 of the schema? Once we merge this any change to the schema will require a version number increment.

Yeah, that should be fine. Pinging @ltrani for schema business and confirmation.

ltrani commented:
> Yeah, that should be fine. Pinging @ltrani for schema business and confirmation.
I can confirm that the schema can be finalize and we can consider this as version 1.0.0
Thank you all for the great job!

krischer commented:
Alright. I'm hitting the big and mighty merge button :-) Thanks everyone!

ThomasLecocq commented:
WOW !! congratulations everyone !!!

megies commented:
This has introduced a test fail on Debian 7 old-stable that wasn't noticed so far because I only now install jsonschema on that docker box.

Can somebody have a look, @ltrani @krischer ?

http://tests.obspy.org/82648/#1

krischer commented:
What is the version of jsonschema on that box?

megies commented:
0.2   https://packages.debian.org/wheezy/python-jsonschema

krischer commented:
> 0.2 https://packages.debian.org/wheezy/python-jsonschema
> 
> (that's ancient, I guess, so probably an raise ImportError('jsonschema version too low') in the code base and a skipIf in the tests should do..)

Dealt with in #1787
